mkosi/gentoo.py

```diff
@@ -0,0 +1,482 @@
+
+        # we use this so we don't need to touch upstream files
+        # we also use this for documenting build environment as much as possible
+
+        emerge_vars_str = ""
+        for k, v in self.emerge_vars.items():
+            emerge_vars_str += f'{k}="$' + "{" + k + "}" + f' {v}"\n'
+
+        ebuild_sh_env_dir.joinpath("mkosi.conf").write_text(
+            dedent(
+                f"""\
+                    # MKOSI: these were used during image creation...
+                    # and some more! see under package.*/
+                    #
+                    # usrmerge (see all under profile/)
+                    {emerge_vars_str}
+                    """
+            )
+        )
+
+    def invoke_emerge(
+        self,
+        args: CommandLineArguments,
+        root: Path,
+        inside_stage3: bool = True,
+        pkgs: List[str] = list(),
+        actions: List[str] = list(),
+        opts: List[str] = list(),
+    ) -> None:
+        if not inside_stage3:
+            from _emerge.main import emerge_main  # type: ignore
+
+            PREFIX_OPTS: List[str] = list()
+            if "--sync" not in actions:
+                PREFIX_OPTS = [
+                    f"--config-root={root.resolve()}",
+                    f"--root={root.resolve()}",
+                    f"--sysroot={root.resolve()}",
+                ]
+
+            emerge_main(pkgs + opts + PREFIX_OPTS + self.emerge_default_opts + actions)
+        else:
+            cmdline = ["/usr/bin/emerge"]
+
+            cmdline.extend(pkgs)
+            cmdline.extend(self.emerge_default_opts)
+            cmdline.extend(opts)
+            cmdline.extend(actions)
+
+            MkosiPrinter.print_step("Invoking emerg(1) inside stage3")
+            run_workspace_command(
+                args,
+                root,
+                cmdline,
+                network=True,
+                env=self.emerge_vars,
+                nspawn_params=self.DEFAULT_NSPAWN_PARAMS + ["--bind=/sys"],
+            )
+
+    def baselayout(self, args: CommandLineArguments, root: Path) -> None:
+        # TOTHINK: sticky bizness when when image profile != host profile
+        self.invoke_emerge(args, root, pkgs=["sys-apps/baselayout"])
+        cmdline = ["env-update"]
+        run_workspace_command(args, root, cmdline, nspawn_params=self.DEFAULT_NSPAWN_PARAMS)
+
+    def update_stage3(self, args: CommandLineArguments, root: Path) -> None:
+        # baselayout gets remerged here because of USE=build which overrides
+        # /etc/os-release; (see below)
+        self.invoke_emerge(args, root, pkgs=self.pkgs_sys, opts=self.EMERGE_UPDATE_OPTS)
+
+        # DONTMOVE: baselayout is settled down by now, dracut needs this for
+        # version_id in /etc/os-release data (UKI)
+        # GENTOO_BUG: https://bugs.gentoo.org/788190
+        with root.joinpath("etc/os-release").open("a") as f:
+            f.write(f"VERSION_ID={args.release}\n")
```

Comment:
`root.joinpath("etc/os-release").write_text(f"VERSION_ID={args.release}\n")`

Reply:
that `open`s the file with `'w'` and resulting image is not acceptable for systemd's bootloader. i couldn't find a way to pass `mode='a'` to `write_text`, will research more.

Reply:
I missed the 'a', sorry. So the current form is OK.

(But that has the problem that that you don't know is VERSION_ID= is already present in the file. So I would expect that you need to filter the file to remove VERSION_ID= anyway. Are you sure it's not there?)

Reply:
# GENTOO_BUG: https://bugs.gentoo.org/788190
i opened that bug report for `VERSION_ID=` missing.